### PR TITLE
Add permissions checks for results controllers [SCI-6071]

### PR DESCRIPTION
**Pull request by aignatov-bio**
Jira ticket: [SCI-6071](https://biosistemika.atlassian.net/browse/SCI-6071)

### What was done
Add permissions checks for results controllers

### CODE REVIEW

File: app/controllers/result_assets_controller.rb [hunk 1]
```diff
@@ -1,10 +1,11 @@
 class ResultAssetsController < ApplicationController
   include ResultsHelper
 
-  before_action :load_vars, only: [:edit, :update, :download]
+  before_action :load_vars, only: [:edit, :update]
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/controllers/result_assets_controller.rb:4]: fix this

File: app/controllers/result_tables_controller.rb [hunk 2]
```diff
@@ -5,8 +5,10 @@ class ResultTablesController < ApplicationController
   before_action :load_vars_nested, only: [:new, :create]
   before_action :convert_contents_to_utf8, only: [:create, :update]
 
-  before_action :check_manage_permissions, only: %i(new create edit update)
+  before_action :check_manage_permissions, only: %i(edit update)
+  before_action :check_create_permissions, only: %i(new create)
   before_action :check_archive_permissions, only: [:update]
+  before_action :check_view_permissions, only: [:download]
```
Comment (okriuchykhin) [inline app/controllers/result_tables_controller.rb:11]: use except for view permissions

File: app/controllers/result_tables_controller.rb [hunk 3]
```diff
@@ -145,8 +147,12 @@ def convert_contents_to_utf8
     end
   end
 
+  def check_create_permissions
+    render_403 unless can_create_my_module_results?(@my_module)
```
Comment (okriuchykhin) [inline app/controllers/result_tables_controller.rb:151]: may be can_create_results? ?

File: app/controllers/result_texts_controller.rb [hunk 4]
```diff
@@ -8,8 +8,10 @@ class ResultTextsController < ApplicationController
   before_action :load_vars, only: [:edit, :update, :download]
   before_action :load_vars_nested, only: [:new, :create]
 
-  before_action :check_manage_permissions, only: %i(new create edit update)
+  before_action :check_manage_permissions, only: %i(edit update)
+  before_action :check_create_permissions, only: %i(new create)
   before_action :check_archive_permissions, only: [:update]
+  before_action :check_view_permissions, only: [:download]
```
Comment (okriuchykhin) [inline app/controllers/result_texts_controller.rb:14]: use except for view permissions

File: app/permissions/result.rb [hunk 5]
```diff
@@ -2,14 +2,39 @@
...
+    result.archived? &&
+      result.unlocked?(result) &&
+      result.my_module.permission_granted?(user, MyModulePermissions::RESULTS_DELETE_ARCHIVED)
+  end
+end
+
+Canaid::Permissions.register_for(Comment) do
+  # Module, its experiment and its project must be active for all the specified
+  # permissions
+  %i(manage_result_comments)
```
Comment (okriuchykhin) [inline app/permissions/result.rb:24]: manage_result_comment

File: app/permissions/result.rb [hunk 6]
```diff
@@ -2,14 +2,39 @@
...
+    can perm do |_, comment|
+      my_module = ::PermissionsUtil.get_comment_module(comment)
+      !my_module.archived_branch?
+    end
+  end
+
+  # module: update/delete comment
+  # result: update/delete comment
+  # step: update/delete comment
+  can :manage_result_comments do |user, comment|
```
Comment (okriuchykhin) [inline app/permissions/result.rb:35]: manage_result_comment

File: app/controllers/result_tables_controller.rb (hunk 3 above)
Comment (aignatov-bio) [reply app/controllers/result_tables_controller.rb:151]: It is part of my_module permissions, we have this format for permissions there - _action__**my_module**__objects_

File: app/permissions/result.rb [hunk 7]
```diff
@@ -2,14 +2,39 @@
...
 
   can :delete_result do |user, result|
-    can_manage_my_module?(user, result.my_module) && result.archived? && result.unlocked?(result)
+    result.archived? &&
+      result.unlocked?(result) &&
+      result.my_module.permission_granted?(user, MyModulePermissions::RESULTS_DELETE_ARCHIVED)
+  end
+end
+
+Canaid::Permissions.register_for(Comment) do
```
Comment (okriuchykhin) [inline app/permissions/result.rb:21]: Change to ResultComment

**Review by okriuchykhin — APPROVED** [review]

**Pull request merged**